### PR TITLE
HIVE-29829: Iceberg: Stream partition column statistics on write, prune them on read

### PR DESCRIPTION
**Pull request by deniskuzZ**
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Per-partition column statistics for Iceberg tables have been reworked end to end, covering how they are stored, written, and read during scans, as well as how readers determine whether the statistics still accurately describe the underlying data.                                                                                                                                                                                                        
  

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Performance driven optimization and design refactor

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description, screenshot and/or a reproducable example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Hive versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
- TestIcebergColStatsFormat
- TestIcebergColStatsWritePolicy
- TestHiveIcebergStatistics

qtests:
- iceberg_colstats_analyze_evolution.q
- iceberg_colstats_folded.q
- iceberg_colstats_granularity.q
- iceberg_colstats_staleness.q
- iceberg_colstats_write_paths.q,
- iceberg_compaction_colstats_compute.q

depends on https://github.com/apache/hive/pull/6711

### CODE REVIEW

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/stats/IcebergColStatsCodec.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+  /**
+   * What a sketch slice is named after the column it belongs to. No column name holds this, so a
+   * blob that carries no sketch simply has no slice named this way.
+   */
+
+  private IcebergColStatsCodec() {
+  }
+
+  /**
+   * Takes the sketch out of the entry, so that what remains is the part every read needs. Returns
```
Comment (Aggarwal-Raghav) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/stats/IcebergColStatsCodec.java]: This javadoc seems misplaced?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/stats/IcebergColStatsCodec.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    return out.toByteArray();
+  }
+
+  /**
+   * The entries the given places name, the rest skipped rather than copied. A scan of a wide table
+   * asks about a few of its columns, and an entry it does not want costs a read nothing beyond the
+   * length it steps over.
+   */
+  static List<byte[]> decodeBlob(ByteBuffer buf, IntPredicate wanted) throws IOException {
+    DataInputStream data = new DataInputStream(new ByteArrayInputStream(ByteBuffers.toByteArray(buf)));
```
Comment (Aggarwal-Raghav) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/stats/IcebergColStatsCodec.java]:
```
DataInputStream data = new DataInputStream(new ByteArrayInputStream(ByteBuffers.toByteArray(buf)));
```
is necessary? the ops like readInt etc can be done directly on ByteBuffer as well.

**Comment by Aggarwal-Raghav** [comment]
This PR is a bit overwhelming, so I’m mostly reviewing to learn and understand the new flow!

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/stats/IcebergColStatsWriter.java [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+      }
+    }
+  }
+
+  @FunctionalInterface
+  private interface ColStatsBlobWriter {
+    void write(PuffinWriter writer) throws IOException, InvalidObjectException;
+  }
+
+  /** Writes one statistics file through the given blobs and commits it for the snapshot. */
```
Comment (Aggarwal-Raghav) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/stats/IcebergColStatsWriter.java]: javadoc fix might be required.

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 4]
```diff
@@ -2130,6 +2130,22 @@ public static enum ConfVars {
         "Whether to use codec pool in ORC. Disable if there are bugs with codec reuse."),
     HIVE_ICEBERG_STATS_SOURCE("hive.iceberg.stats.source", "iceberg",
         "Use stats from iceberg table snapshot for query planning. This has two values metastore and iceberg"),
+    HIVE_ICEBERG_STATS_COLLECT_PART_LEVEL("hive.iceberg.stats.collect.partlevel", false,
```
Comment (difin) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:2133]: Why should it be false by default?

Comment (deniskuzZ) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:2133]: need to be enabled with caution: on wide tables might not justify the purpose, table-level is recommended

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/stats/IcebergColStatsCodec.java (hunk 2 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/stats/IcebergColStatsCodec.java]: good point, note ByteBuffers.toByteArray(buf) copy was inherited from master

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/stats/IcebergColStatsWriter.java [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.ByteBuffers;
+import org.apache.iceberg.util.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Writes the column statistics of one gather as the table's statistics file, per the policy the
```
Comment (difin) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/stats/IcebergColStatsWriter.java]: "one gather" is not very understandable

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java [hunk 6]
```diff
@@ -804,32 +923,128 @@ public static TransformSpec getTransformSpec(Table table, String transformName,
...
+  /**
+   * The partitions written since a snapshot. Null when the walk cannot answer - an expired
+   * snapshot broke the chain, or the bound was reached - and every partition then counts as
+   * changed. The bound is on manifests read, not on snapshots walked.
+   */
+  static Set<String> partitionsChangedSince(Table table, Snapshot snapshot, long sinceSnapshotId,
+      Configuration conf, boolean capped) {
+    // the walk reads manifests, and every column of a query asks the same question of them
+    String cacheKey = MODIFIED_PARTITIONS_PREFIX + table.name() + '.' + snapshot.snapshotId() + '.' +
+        sinceSnapshotId + '.' + capped;
```
Comment (difin) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: Should this be converted to `String.format()` for better readability?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/stats/TestIcebergColStatsFormat.java [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+  public void blobsLyingCloseTogetherAreTakenInOneRead() throws Exception {
+    List<byte[]> blobs = org.apache.iceberg.relocated.com.google.common.collect.Lists.newArrayList();
+    for (int i = 0; i < 3; i++) {
+      blobs.add(IcebergColStatsCodec.encodeBlob(
+          List.of(IcebergColStatsCodec.encodeEntry(longColumn("c" + i, i))), List.of(1)));
+    }
+    // laid end to end, so no gap is worth a second request
+    List<Long> offsets = List.of(0L, (long) blobs.get(0).length,
+        (long) blobs.get(0).length + blobs.get(1).length);
+    List<BlobMetadata> meta = org.apache.iceberg.relocated.com.google.common.collect.Lists.newArrayList();
```
Comment (difin) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/stats/TestIcebergColStatsFormat.java]: This would have a better readability if the class was imported: `org.apache.iceberg.relocated.com.google.common.collect.Lists`

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java [hunk 8]
```diff
@@ -291,6 +294,7 @@ default List<ColumnStatisticsObj> getColStatistics(org.apache.hadoop.hive.ql.met
     return null;
   }
 
+
```
Comment (difin) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java]: nit: extra new line

**Review by difin — APPROVED** [review]
LGTM +1

I left a few comments.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/stats/IcebergColStatsReader.java [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+        BlobMetadata held = ordered.get(last);
+        long gap = ordered.get(next).offset() - (held.offset() + held.length());
+        long span = ordered.get(next).offset() + ordered.get(next).length() - ordered.get(first).offset();
+        if (gap > minSeek || span > maxReadSize) {
+          break;
+        }
+        last = next;
+      }
+      long start = ordered.get(first).offset();
+      int length = (int) (ordered.get(last).offset() + ordered.get(last).length() - start);
```
Comment (difin) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/stats/IcebergColStatsReader.java]: `length` is computed as long, but stored as int. Will it be sufficient?

Should we assert `length <= maxReadSize` explicitly before allocating?

Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/stats/IcebergColStatsReader.java]: switched to Math.toIntExact — same idiom as Iceberg's own PuffinReader. It has to be an int either way (byte[] and IOUtil.readFully are both int-bound)

**Comment by difin** [comment]
PR description lists `TestIcebergManifestColStats`, but I don't see this class in the diff.

**Comment by difin** [comment]
There are 43 Sonar issues, some looks actionable, like replacing String concatenation blocks with `String.format()`, adding try-with-resources.

**Comment by deniskuzZ** [comment]
> PR description lists `TestIcebergManifestColStats`, but I don't see this class in the diff.

removed, it was extracted into separate PR
